Readme.md
![Apex Serverless Architecture](assets/title.png)

Apex lets you build, deploy, and manage [AWS Lambda](https://aws.amazon.com/lambda/) functions with ease. With Apex you can use languages that are not natively supported by AWS Lambda through the use of a Node.js shim injected into the build. A variety of workflow related tooling is provided for testing functions, rolling back deploys, viewing metrics, tailing logs, hooking into the build system and more.

This project is designed for event-driven pipelines as it does not abstract away FaaS (functions as a service). If you are building web applications, APIs, or sites, consider using [Apex Up](https://github.com/apex/up), which provides a more out-of-the-box experience for these use-cases.

## Installation

On macOS, Linux, or OpenBSD run the following:

```
curl https://raw.githubusercontent.com/apex/apex/master/install.sh | sh
```

Note that you may need to run the `sudo` version below, or alternatively chown `/usr/local`:
```
curl https://raw.githubusercontent.com/apex/apex/master/install.sh | sudo sh
```

On Windows download [binary](https://github.com/apex/apex/releases).

After downloading, rename binary file 'apex.exe', then add to PATH.

If already installed, upgrade with:

```
apex upgrade
```

## Runtimes

Currently supports:

- Node.js
- Golang
- Python
- Java
- Rust
- Clojure

Example projects for all supported runtimes can be found in [_examples](_examples) directory.

## Features

- Supports languages Lambda does not natively support via shim
- Binary install (install apex quickly for continuous deployment in CI etc)
- Hook support for running commands (transpile code, lint, dependency management , etc)
- Batteries included but optional (opt-in to higher level abstractions)
- Environment variable population via command-line, file, or inline config
- Idempotent deployments (checksums skip already-deployed code)
- Multiple environments via `project.ENV.json` and `function.ENV.json` files
- Configuration inheritance and overrides
- Command-line function invocation with JSON streams
- Command & function name autocompletion
- Function name globbing (ex: `apex deploy api_*`)
- Transparently generates a zip for your deploy
- Project bootstrapping with optional Terraform support
- Function metrics and cost analysis
- Ignore deploying files with .apexignore
- Function rollback support
- Tail function logs
- Concurrency for quick deploys
- Dry-run to preview changes
- VPC support

## Sponsors

Does your company use Apex? Help keep the project bug-free and feature rich by [sponsoring the project](https://opencollective.com/apex#sponsor).

  <a href="https://opencollective.com/apex/sponsors/0/website" target="_blank"><img src="https://opencollective.com/apex/sponsors/0/avatar"></a>
  <a href="https://opencollective.com/apex/sponsors/1/website" target="_blank"><img src="https://opencollective.com/apex/sponsors/1/avatar"></a>
  <a href="https://opencollective.com/apex/sponsors/2/website" target="_blank"><img src="https://opencollective.com/apex/sponsors/2/avatar"></a>
  <a href="https://opencollective.com/apex/sponsors/3/website" target="_blank"><img src="https://opencollective.com/apex/sponsors/3/avatar"></a>
  <a href="https://opencollective.com/apex/sponsors/4/website" target="_blank"><img src="https://opencollective.com/apex/sponsors/4/avatar"></a>

## Backers

Love our work and community? [Become a backer](https://opencollective.com/apex).

  <a href="https://opencollective.com/apex/backers/0/website" target="_blank"><img src="https://opencollective.com/apex/backers/0/avatar"></a>
  <a href="https://opencollective.com/apex/backers/1/website" target="_blank"><img src="https://opencollective.com/apex/backers/1/avatar"></a>
  <a href="https://opencollective.com/apex/backers/2/website" target="_blank"><img src="https://opencollective.com/apex/backers/2/avatar"></a>
  <a href="https://opencollective.com/apex/backers/3/website" target="_blank"><img src="https://opencollective.com/apex/backers/3/avatar"></a>
  <a href="https://opencollective.com/apex/backers/4/website" target="_blank"><img src="https://opencollective.com/apex/backers/4/avatar"></a>
  <a href="https://opencollective.com/apex/backers/5/website" target="_blank"><img src="https://opencollective.com/apex/backers/5/avatar"></a>
  <a href="https://opencollective.com/apex/backers/6/website" target="_blank"><img src="https://opencollective.com/apex/backers/6/avatar"></a>
  <a href="https://opencollective.com/apex/backers/7/website" target="_blank"><img src="https://opencollective.com/apex/backers/7/avatar"></a>
  <a href="https://opencollective.com/apex/backers/8/website" target="_blank"><img src="https://opencollective.com/apex/backers/8/avatar"></a>
  <a href="https://opencollective.com/apex/backers/9/website" target="_blank"><img src="https://opencollective.com/apex/backers/9/avatar"></a>
  <a href="https://opencollective.com/apex/backers/10/website" target="_blank"><img src="https://opencollective.com/apex/backers/10/avatar"></a>
  <a href="https://opencollective.com/apex/backers/11/website" target="_blank"><img src="https://opencollective.com/apex/backers/11/avatar"></a>
  <a href="https://opencollective.com/apex/backers/12/website" target="_blank"><img src="https://opencollective.com/apex/backers/12/avatar"></a>
  <a href="https://opencollective.com/apex/backers/13/website" target="_blank"><img src="https://opencollective.com/apex/backers/13/avatar"></a>
  <a href="https://opencollective.com/apex/backers/14/website" target="_blank"><img src="https://opencollective.com/apex/backers/14/avatar"></a>
  <a href="https://opencollective.com/apex/backers/15/website" target="_blank"><img src="https://opencollective.com/apex/backers/15/avatar"></a>
  <a href="https://opencollective.com/apex/backers/16/website" target="_blank"><img src="https://opencollective.com/apex/backers/16/avatar"></a>
  <a href="https://opencollective.com/apex/backers/17/website" target="_blank"><img src="https://opencollective.com/apex/backers/17/avatar"></a>
  <a href="https://opencollective.com/apex/backers/18/website" target="_blank"><img src="https://opencollective.com/apex/backers/18/avatar"></a>
  <a href="https://opencollective.com/apex/backers/19/website" target="_blank"><img src="https://opencollective.com/apex/backers/19/avatar"></a>
  <a href="https://opencollective.com/apex/backers/20/website" target="_blank"><img src="https://opencollective.com/apex/backers/20/avatar"></a>
  <a href="https://opencollective.com/apex/backers/21/website" target="_blank"><img src="https://opencollective.com/apex/backers/21/avatar"></a>
  <a href="https://opencollective.com/apex/backers/22/website" target="_blank"><img src="https://opencollective.com/apex/backers/22/avatar"></a>
  <a href="https://opencollective.com/apex/backers/23/website" target="_blank"><img src="https://opencollective.com/apex/backers/23/avatar"></a>
  <a href="https://opencollective.com/apex/backers/24/website" target="_blank"><img src="https://opencollective.com/apex/backers/24/avatar"></a>
  <a href="https://opencollective.com/apex/backers/25/website" target="_blank"><img src="https://opencollective.com/apex/backers/25/avatar"></a>
  <a href="https://opencollective.com/apex/backers/26/website" target="_blank"><img src="https://opencollective.com/apex/backers/26/avatar"></a>
  <a href="https://opencollective.com/apex/backers/27/website" target="_blank"><img src="https://opencollective.com/apex/backers/27/avatar"></a>
  <a href="https://opencollective.com/apex/backers/28/website" target="_blank"><img src="https://opencollective.com/apex/backers/28/avatar"></a>
  <a href="https://opencollective.com/apex/backers/29/website" target="_blank"><img src="https://opencollective.com/apex/backers/29/avatar"></a>

## Example

Apex projects are made up of a `project.json` configuration file, and zero or more Lambda functions defined in the "functions" directory. Here's an example file structure:

```
project.json
functions
├── bar
│   ├── function.json
│   └── index.js
└── foo
    ├── function.json
    └── index.js
```

The `project.json` file defines project level configuration that applies to all functions, and defines dependencies. For this simple example the following will do:

```json
{
  "name": "example",
  "description": "Example project"
}
```

Each function uses a `function.json` configuration file to define function-specific properties such as the runtime, amount of memory allocated, and timeout. This file is completely optional, as you can specify defaults in your `project.json` file. For example:

```json
{
  "name": "bar",
  "description": "Node.js example function",
  "runtime": "nodejs4.3",
  "memory": 128,
  "timeout": 5,
  "role": "arn:aws:iam::293503197324:role/lambda"
}
```

Now the directory structure for your project would be:

```
project.json
functions
├── bar
│   └── index.js
└── foo
    └── index.js
```

Finally the source for the functions themselves look like this in Node.js:

```js
console.log('start bar')
exports.handle = function(e, ctx) {
  ctx.succeed({ hello: e.name })
}
```

Apex operates at the project level, but many commands allow you to specify specific functions. For example you may deploy the entire project with a single command:

```
$ apex deploy
```

Or whitelist functions to deploy:

```
$ apex deploy foo bar
```

Invoke it!

```
$ echo '{ "name": "Tobi" }' | apex invoke bar
{ "hello": "Tobi" }
```

See the [Documentation](http://apex.run) for more information.

## Links

- [Website](http://apex.run)
- [Medium](https://medium.com/@tjholowaychuk)
- [Slack](https://chat.apex.sh/)

---

[![Build Status](https://semaphoreci.com/api/v1/projects/d27ff350-b9c5-4d99-96e5-64b1afb441c5/649392/badge.svg)](https://semaphoreci.com/tj/apex)
[![Slack Status](https://apex-slackin.herokuapp.com/badge.svg)](https://apex-slackin.herokuapp.com/)
[![GoDoc](https://godoc.org/github.com/apex/apex?status.svg)](https://godoc.org/github.com/apex/apex)
![](https://img.shields.io/badge/license-MIT-blue.svg)
![](https://img.shields.io/badge/status-stable-green.svg) [![OpenCollective](https://opencollective.com/apex/backers/badge.svg)](#backers) [![OpenCollective](https://opencollective.com/apex/sponsors/badge.svg)](#sponsors)

<a href="https://apex.sh"><img src="http://tjholowaychuk.com:6000/svg/sponsor"></a>


History.md

v1.0.0-rc4 / 2018-08-29
=======================

  * add `region` field to project configuration (#889)
  * update AWS SDK versions for assuming roles. Closes #886

v1.0.0-rc3 / 2018-05-07
=======================

  * add AWS endpoint flag. Closes #794 (#875)
  * fix support for >50 lambda versions. Closes https://github.com/apex/apex/issues/846 (#882)
  * fix broken Slack Status link on website (#880)

v1.0.0-rc2 / 2018-01-17
=======================

  * add new upgrade mechanism to match Up
  * add utils from Up which are necessary for upgrade
  * add native Go support. Closes #864

v0.16.0 / 2017-10-14
====================

  * add `./vendor`
  * refactor: move `process.env` out of the shim event handlers
  * refactor shim, allowing for concurrent calls to handler (#822)
  * remove analytics. Closes #777
  * fix node runtime
  * fix deploying java and clojure functions from zipfile (#820)
  * fix #815 - clojure deployments being doubled by java plugin (#816)
  * fix for issue #380 (#814)
  * fix: move -C chdir flag check to beginning of preparation function. (#766)

v0.15.0 / 2017-06-11
====================

  * add passing of maps that describe all apex functions to terraform Closes #737 (#744)
  * added nodejs4.3-edge runtime (#741)
  * add ListFunctions to min IAM policy example (#733)
  * add documentation to set alias to another alias (#732)
  * add arbitrary python version support. Closes #723
  * Add lambda:ListAliases to min IAM policy (#729)
  * add --alias docs
  * change DefaultRetainedVersions to 25
  * Update function deploy to update the specified alias even if code has not changed. Closes #686 (#712)
  * upgrade example to use nodejs version 6.10 (#719)
  * fix tabs in deploy example

v0.14.0 / 2017-03-31
====================

  * add support for deploying existing zips. Closes #480
  * change index.js inference to use v6.10
  * fix empty DeadLetterARN config triggering deploy no matter what. Possibly Closes #701 (#713)
  * fix webpack2 example for version 2.3.1 (#709)

v0.13.1 / 2017-03-23
====================

  * add Node v6.10 support. Closes #493
  * change runtime to nodejs43 for rust lambdas (#689)
  * fix context for gh ListReleases call (#691)
  * fix alias command example by removing v prefix: v5 should be just 5 (#681)

v0.13.0 / 2017-02-15
====================

  * add clojure support (#674)
  * add variable apex_function_NAME_name to the exported variable to terraform (Closes #654) (#660)
  * add support for DLQ ARN (#649)
  * add support for KMS ARN (#648)
  * add alias command (#647)

v0.12.0 / 2017-01-06
====================

  * add `exec` sub-command for pass-through of env vars. Closes #619
  * add improved message if curl failed due to permission error (#634)
  * add initial rust support (#549)
  * add function arn to list output (#627)
  * add IAM role support (#614)
  * fix shim runtime for new functions, now "nodejs4.3"

v0.11.0 / 2016-11-21
====================

  * add comparison of environment when performing change check
  * add support for APEX_ENVIRONMENT=prod apex deploy (#586)
  * add a minimum IAM policy to the documentation. (#529)
  * remove env hack, replace with native env support (#600)
  * fix overrides of VPC in function.json. Closes #597 (#598)
  * fix .apexignore fails when folder without trailling slash (#550)

v0.10.3 / 2016-09-05
====================

* add function.json fallback for any env if it is present. Closes #471
* add aliases to apex list. Closes #430 (#457)
* fix config comparison function. Closes #528 (#530)
* fix error handling for incorrect 'handler' definition for python (fixes #498) (#501)
* fix: use executable's path as the temp path (#488)
* fix: run hooks before runtime plugins to fix (#491) (#492)
* fix ignoring of function.json when defaultEnvironment is used. Closes #467
* fix for go-github api changes. Closes #463 (#465)
* fix custom clean hook was not working on golang runtime. Closes #461 (#464)

v0.10.2 / 2016-06-16
====================

  * fix version for upgrades

v0.10.1 / 2016-06-14
====================

  * add missing --env-file for build command. Closes #452

v0.10.0 / 2016-06-14
====================

  * add getting-started docs with init command
  * add infra docs
  * add role bootstrapping to `apex init` for a smoother experience
  * add project.json config "profile" support
  * add `--env-file` support. Closes #387
  * add multiple deployment targets for different environments (#432)
  * add missing `-s, --set` environment support to the build command. Closes #447
  * add retainedVersions as a pointer, letting you zero. Closes #407
  * refactor project init
  * remove api-gateway example since it is incomplete. Closes #445
  * fix flags before infra sub-command. Closes #421
  * fix panic when 'apex invoke -L' throws an error (#431)
  * fix setting handler func in Python runtime

v0.9.0 / 2016-05-10
===================

  * add symlink docs
  * add autocomplete docs
  * add dynamic auto-completion (#403)
  * add function globbing (#397)
  * add basic Segment analytics
  * add support for symlinks. Closes #320 (#375)
  * add marking not deployed functions on list (#374)
  * add multiple funcs rollback. Closes #372 (#373)
  * add support for updating function runtimes (apex deploy). Closes #369
  * add metric price (#368)
  * add implicit APEX_FUNCTION_NAME and LAMBDA_FUNCTION_NAME env vars
  * add apex environment as tfvar `apex_environment`
  * fix symlink support. Closes #401 #288 (#404)
  * fix install script, parse Github tag JSON with no newlines fixes #400 (#402)
  * fix missing logs. Closes #393
  * fix new lines in Hooks docs. Closes #390
  * fix generated boilerplate package
  * fix panic when trying to set incomplete -s var (#379)
  * fix infra command terraform flags pass-through. Closes #325 (#386)
  * fix missing defaultEnvironment in project.json. Closes #382

v0.8.0 / 2016-04-08
===================

  * add nodejs 4.3.2 support. Closes #356

boot/boot.go
package boot

import (
	"fmt"
	"io/ioutil"
	"os"
	"strings"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/iam"
	"github.com/aws/aws-sdk-go/service/iam/iamiface"
	"github.com/tj/go-prompt"

	"github.com/apex/apex/boot/boilerplate"
)

// TODO(tj): attempt creation of S3 bucket to streamline that as well
// TODO(tj): idempotency, if the project exists then skip all this, or provision AWS
// TODO(tj): validate the name, \w+ or similar should be fine

var logo = `

             _    ____  _______  __
            / \  |  _ \| ____\ \/ /
           / _ \ | |_) |  _|  \  /
          / ___ \|  __/| |___ /  \
         /_/   \_\_|   |_____/_/\_\

`

var projectConfig = `{
  "name": "%s",
  "description": "%s",
  "memory": 128,
  "timeout": 5,
  "role": "%s",
  "environment": {}
}`

var setupCompleteVanilla = `
  Setup complete, deploy those functions!

    $ apex deploy
`

var iamAssumeRolePolicy = `{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": "lambda.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}`

var iamLogsPolicy = `{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": [
        "logs:*"
      ],
      "Effect": "Allow",
      "Resource": "*"
    }
  ]
}`

// Bootstrapper initializes a project and AWS account for the user.
type Bootstrapper struct {
	IAM    iamiface.IAMAPI
	Region string

	name        string
	description string
}

// Boot the project.
func (b *Bootstrapper) Boot() error {
	fmt.Println(logo)

	if b.isProject() {
		help("I've detected a ./project.json file, this seems to already be a project!")
		return nil
	}

	help("Enter the name of your project. It should be machine-friendly, as this\nis used to prefix your functions in Lambda.")
	b.name = prompt.StringRequired(indent("  Project name: "))

	help("Enter an optional description of your project.")
	b.description = prompt.String(indent("  Project description: "))

	fmt.Println()
	return b.bootVanilla()
}

// check if there's a project.
func (b *Bootstrapper) isProject() bool {
	_, err := os.Stat("project.json")
	return err == nil
}

// Bootstrap without Terraform.
func (b *Bootstrapper) bootVanilla() error {
	iamRole, err := b.createRole()
	if err != nil {
		return err
	}

	if err := b.initProjectFiles(iamRole); err != nil {
		return err
	}

	fmt.Println(setupCompleteVanilla)
	return nil
}

// Create IAM role, returning the ARN.
func (b *Bootstrapper) createRole() (string, error) {
	roleName := fmt.Sprintf("%s_lambda_function", b.name)
	policyName := fmt.Sprintf("%s_lambda_logs", b.name)

	logf("creating IAM %s role", roleName)
	role, err := b.IAM.CreateRole(&iam.CreateRoleInput{
		RoleName:                 &roleName,
		AssumeRolePolicyDocument: aws.String(iamAssumeRolePolicy),
	})

	if err != nil {
		return "", fmt.Errorf("creating role: %s", err)
	}

	logf("creating IAM %s policy", policyName)
	policy, err := b.IAM.CreatePolicy(&iam.CreatePolicyInput{
		PolicyName:     &policyName,
		Description:    aws.String("Allow lambda_function to utilize CloudWatchLogs. Created by apex(1)."),
		PolicyDocument: aws.String(iamLogsPolicy),
	})

	if err != nil {
		return "", fmt.Errorf("creating policy: %s", err)
	}

	logf("attaching policy to lambda_function role.")
	_, err = b.IAM.AttachRolePolicy(&iam.AttachRolePolicyInput{
		RoleName:  &roleName,
		PolicyArn: policy.Policy.Arn,
	})

	if err != nil {
		return "", fmt.Errorf("creating policy: %s", err)
	}

	return *role.Role.Arn, nil
}

// Initialize project files such as project.json and ./functions.
func (b *Bootstrapper) initProjectFiles(iamRole string) error {
	logf("creating ./project.json")

	project := fmt.Sprintf(projectConfig, b.name, b.description, iamRole)

	if err := ioutil.WriteFile("project.json", []byte(project), 0644); err != nil {
		return err
	}

	logf("creating ./functions")
	return boilerplate.RestoreAssets(".", "functions")
}

// help string output.
func help(s string) {
	os.Stdout.WriteString("\n")
	os.Stdout.WriteString(indent(s))
	os.Stdout.WriteString("\n\n")
}

// indent multiline string with 2 spaces.
func indent(s string) (out string) {
	for _, l := range strings.SplitAfter(s, "\n") {
		out += fmt.Sprintf("  %s", l)
	}

	return
}

// logf outputs a log message.
func logf(s string, v ...interface{}) {
	fmt.Printf("  \033[34m[+]\033[0m %s\n", fmt.Sprintf(s, v...))
}
